### ISSUE TITLE
Can't build JupyterHub-ready images with runtime.txt specifying python-2.7

### ISSUE THREAD

**Issue by mgerring** [jupyter/repo2docker#314]
I have a [Github repository](https://github.com/econ-ark/DemARK) containing notebooks that need to run in Python 2.7. I am trying to deploy JupyterHub on AWS with an image created from that repository.

According to the repo2docker documentation, I need to include a `runtime.txt` file specifying python 2.7. I also need to include `jupyterhub` in the `requirements.txt` file. If I include both, the build fails, because `repo2docker` attempts to install `jupyterhub` with Python 2.7.

Any ideas on how to fix this?

**Comment by choldgraf** [comment]
hmmm - you shouldn't need to manually install `jupyterhub` with requirements.txt, can you try without it? (e.g. [see this example](http://repo2docker.readthedocs.io/en/latest/samples.html#system-specifying-runtime-environments))

**Comment by mgerring** [comment]
The documentation for Jupyterhub says you need to explicitly list jupyterhub in the requirements. I’ll give it a try though.

**Comment by choldgraf** [comment]
could you send a link to the page you're referring to? Are you talking about the zero to jupyterhub guide for kubernetes?

**Comment by betatim** [comment]
In the [repo2docker section of z2jh](https://zero-to-jupyterhub.readthedocs.io/en/latest/user-environment.html#build-a-custom-docker-image-with-repo2docker) it does say you should add JupyterHub to the `requirements.txt`. This change was [added recently](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/689).

**Comment by betatim** [comment]
Just tried to build a vanilla image with `repo2docker` and use it with a local JupyterHub. The files are here: https://gist.github.com/betatim/2ffd92a202dd8cd3a3728968e2bc879a The problem I encounter with this is that `jupyterhub-singleuser` isn't found on the PATH. Using different runtimes doesn't have any effect. (You also have to explicitly set the command on the DockerSpawner config not `c.Spawner.cmd`.)

**Comment by mgerring** [comment]
OK. Not sure what to do with this information -- my configuration is in a YAML file and I'm not totally sure what your config.py is doing or how to use it in my deployment.

**Comment by choldgraf** [comment]
actually now that I think about it, I believe that repo2docker will let you specify a `requirements3.txt` file if you are building a python 2 environment. Whatever is in that file should be installed using Python 3, so try putting `jupyterhub` in that one and see if that works @mgerring 

http://mybinder.readthedocs.io/en/latest/sample_repos.html?highlight=requirements3.txt#specifying-a-python-2-environment-with-runtime-txt

**Comment by betatim** [comment]
Does Chris' suggestion work for you @mgerring?

**Comment by mgerring** [comment]
No, it doesn't. I get the following error running repo2docker:

```
Traceback (most recent call last):
  File "/Users/matthew/anaconda3/bin/jupyter-repo2docker", line 11, in <module>
    load_entry_point('jupyter-repo2docker', 'console_scripts', 'jupyter-repo2docker')()
  File "/Users/matthew/dev/repo2docker/repo2docker/__main__.py", line 7, in main
    f.start()
  File "/Users/matthew/dev/repo2docker/repo2docker/app.py", line 603, in start
    self.log.debug(picked_buildpack.render(),
  File "/Users/matthew/dev/repo2docker/repo2docker/buildpacks/base.py", line 310, in render
    for user, script in self.get_assemble_scripts():
  File "/Users/matthew/dev/repo2docker/repo2docker/buildpacks/python/__init__.py", line 56, in get_assemble_scripts
    '${NB_PYTHON_PREFIX}/bin/pip install --no-cache-dir -r "{}"'.format(nb_requirements_file)
KeyError: 'NB_PYTHON_PREFIX'
```

**Comment by betatim** [comment]
I can't reproduce this. Running `repo2docker --debug https://github.com/binder-examples/python2_runtime` builds a working image where the default Python version of notebooks is Python 2.

**Comment by choldgraf** [comment]
@mgerring can you confirm that you're running the latest repo2docker? Try running from "master"

**Comment by betatim** [comment]
Closing this, let's reopen it if we can reproduce it or there is other activity.

**Comment by betatim** [comment]
@mgerring this might be fixed by #377 if you want to give that a spin.
